narwhal/consensus/src/consensus.rs

```diff
@@ -60,37 +60,44 @@ impl ConsensusState {
     }
 
     #[instrument(level = "info", skip_all)]
-    pub async fn new_from_store(
+    pub fn new_from_store(
         genesis: Vec<Certificate>,
         metrics: Arc<ConsensusMetrics>,
         recover_last_committed: HashMap<PublicKey, Round>,
         cert_store: CertificateStore,
         gc_depth: Round,
-    ) -> Self {
+    ) -> (Self, bool) {
+        // We return a bool here which is always true to use as a "recovery token". This
+        // allows us to ensure that the primary is spawned only after the
+        // consensus is guaranteed to be in a state where it can process messages it
+        // receives from the primary when the primary starts up. We do this by passing the
+        // recovery token generated here and checking it before the primary spawn method.
         let last_committed_round = *recover_last_committed
             .iter()
             .max_by(|a, b| a.1.cmp(b.1))
             .map(|(_k, v)| v)
             .unwrap_or_else(|| &0);
 
         if last_committed_round == 0 {
-            return Self::new(genesis, metrics);
+            return (Self::new(genesis, metrics), true);
```

Comment:
If the `bool` response is always `true` I am quite skeptical on what's the value of this. Unless I am missing something here, I believe we could just omit completely this and let the `new_from_store` do its thing - whenever it returns we do have the guarantee that the previous state has been successfully restored (if any exists). The only reason for this to fail is for the method to panic

Reply:
I think you may be right, since we took the set up out of the async block, when the encapsulating function returns, it would be complete.